filters/openpolicyagent/openpolicyagent.go

```diff
@@ -476,38 +477,77 @@ func (registry *OpenPolicyAgentRegistry) new(store storage.Store, configBytes []
 // Start asynchronously starts the policy engine's plugins that download
 // policies, report status, etc.
 func (opa *OpenPolicyAgentInstance) Start(ctx context.Context, timeout time.Duration) error {
-	err := opa.manager.Start(ctx)
 
-	if err != nil {
-		return err
-	}
+	discoveryPlugin := discovery.Lookup(opa.manager)
 
-	// check readiness of all plugins
-	pluginsReady := func() bool {
-		for _, status := range opa.manager.PluginStatus() {
-			if status != nil && status.State != plugins.StateOK {
-				return false
+	done := make(chan struct{})
+	failed := make(chan error, 1)
+
+	var registerDiscoveryListenerOnce sync.Once
+	registerDiscoveryListenerOnce.Do(func() {
+		discoveryPlugin.RegisterListener("startuplistener", func(status bundle.Status) {
+			handleStatusErrors(status, failed, "discovery plugin")
+		})
+	})
+
+	// Register listener for "bundle" status
+	var registerBundleListenerOnce sync.Once
+	opa.manager.RegisterPluginStatusListener("bundlelistener", func(status map[string]*plugins.Status) {
+		if _, exists := status["bundle"]; exists {
+			bundlePlugin := bundle.Lookup(opa.manager)
+			if bundlePlugin != nil {
+				registerBundleListenerOnce.Do(func() {
+					bundlePlugin.Register("startuplistener", func(status bundle.Status) {
+						handleStatusErrors(status, failed, "bundle plugin")
+					})
+				})
 			}
 		}
-		return true
-	}
+	})
+	defer opa.manager.UnregisterPluginStatusListener("bundlelistener")
 
-	err = waitFunc(ctx, pluginsReady, 100*time.Millisecond)
+	// Register listener for general plugin status checks
+	opa.manager.RegisterPluginStatusListener("generalStartUpListener", func(status map[string]*plugins.Status) {
+		for _, pluginStatus := range status {
+			if pluginStatus != nil && pluginStatus.State != plugins.StateOK {
+				return
+			}
+		}
+		done <- struct{}{}
```

Comment:
done channel is unbuffered and this will block until another thread reads from it.
For signalling it might be better to `close(done)`.